spacy/ml/models/coref.py

```diff
@@ -0,0 +1,595 @@
+    """
+    if not clusters:
+        return []
+
+    xp = span_predictor.ops.xp
+    heads_ids = xp.asarray(sorted(i for cluster in clusters for i in cluster))
+    scores = span_predictor.predict((sent_ids, words, heads_ids))
+    starts = scores[:, :, 0].argmax(axis=1).tolist()
+    ends = (scores[:, :, 1].argmax(axis=1) + 1).tolist()
+
+    head2span = {
+        head: (start, end) for head, start, end in zip(heads_ids.tolist(), starts, ends)
+    }
+
+    return [[head2span[head] for head in cluster] for cluster in clusters]
+
+
+# TODO add docstring for this, maybe move to utils.
+# This might belong in the component.
+def _clusterize(model, scores: Floats2d, top_indices: Ints2d):
+    xp = model.ops.xp
+    antecedents = scores.argmax(axis=1) - 1
+    not_dummy = antecedents >= 0
+    coref_span_heads = xp.arange(0, len(scores))[not_dummy]
+    antecedents = top_indices[coref_span_heads, antecedents[not_dummy]]
+    n_words = scores.shape[0]
+    nodes = [GraphNode(i) for i in range(n_words)]
+    for i, j in zip(coref_span_heads.tolist(), antecedents.tolist()):
+        nodes[i].link(nodes[j])
+        assert nodes[i] is not nodes[j]
+
+    clusters = []
+    for node in nodes:
+        if len(node.links) > 0 and not node.visited:
+            cluster = []
+            stack = [node]
+            while stack:
+                current_node = stack.pop()
+                current_node.visited = True
+                cluster.append(current_node.id)
+                stack.extend(link for link in current_node.links if not link.visited)
+            assert len(cluster) > 1
+            clusters.append(sorted(cluster))
+    return sorted(clusters)
+
+
+def build_get_head_metadata(prefix):
+    # TODO this name is awful, fix it
+    model = Model(
+        "HeadDataProvider", attrs={"prefix": prefix}, forward=head_data_forward
+    )
+    return model
+
+
+def head_data_forward(model, docs, is_train):
+    """A layer to generate the extra data needed for the span predictor."""
+    sent_ids = []
+    head_ids = []
+    prefix = model.attrs["prefix"]
+    for doc in docs:
+        sids = model.ops.asarray2i(get_sentence_ids(doc))
+        sent_ids.append(sids)
+        heads = []
+        for key, sg in doc.spans.items():
+            if not key.startswith(prefix):
+                continue
+            for span in sg:
+                # TODO warn if spans are more than one token
+                heads.append(span[0].i)
+        heads = model.ops.asarray2i(heads)
+        head_ids.append(heads)
+    # each of these is a list with one entry per doc
+    # backprop is just a placeholder
+    # TODO it would probably be better to have a list of tuples than two lists of arrays
+    return (sent_ids, head_ids), lambda x: []
+
+
+class CorefScorer(torch.nn.Module):
+    """Combines all coref modules together to find coreferent spans.
+
+    Attributes:
+        epochs_trained (int): number of epochs the model has been trained for
```

Comment:
The two references to `epochs_trained` in the doc strings here can be cleaned up